packages/ckeditor5-engine/src/model/utils/insertcontent.ts

```diff
@@ -77,18 +78,111 @@ export default function insertContent(
 		}
 
 		const insertion = new Insertion( model, writer, selection.anchor! );
+		const markersPosition = [];
+		const fakeMarkerElements = [];
 
 		let nodesToInsert: any;
 
 		if ( content.is( 'documentFragment' ) ) {
+			if ( content.markers.size ) {
+				for ( const [ name, range ] of content.markers ) {
+					const { start, end } = range;
+					const isCollapsed = start.isEqual( end );
+
+					markersPosition.push( {
+						position: start,
+						name,
+						isCollapsed
+					}, {
+						position: end,
+						name,
+						isCollapsed
+					} );
+				}
+
+				markersPosition.sort( ( { position: posA }, { position: posB } ) => posA.isBefore( posB ) ? 1 : -1 );
+
+				for ( const { position, name, isCollapsed } of markersPosition ) {
+					let fakeElement = null;
+					let collapsed = null;
+					const isBeginning = position.parent === content && position.isAtStart;
+					const isEnding = position.parent === content && position.isAtEnd;
+
+					if ( !isBeginning && !isEnding ) {
+						fakeElement = writer.createElement( '$marker', { 'data-name': name } );
+						writer.insert( fakeElement, position );
+					} else if ( isCollapsed ) {
+						collapsed = isBeginning ? 'start' as const : 'end' as const;
+					}
+
+					fakeMarkerElements.push( {
+						name,
+						element: fakeElement,
+						collapsed
+					} );
+				}
+			}
+
 			nodesToInsert = content.getChildren();
 		} else {
 			nodesToInsert = [ content ];
 		}
 
 		insertion.handleNodes( nodesToInsert );
 
-		const newRange = insertion.getSelectionRange();
+		let newRange = insertion.getSelectionRange();
+
+		if ( content.is( 'documentFragment' ) && fakeMarkerElements.length ) {
+			const selectionLiveRange = newRange ? LiveRange.fromRange( newRange ) : null;
+			const markersData: Record<string, Array<Position>> = {};
```

Comment:
I'd rather use `Position[]` instead of `Array<Position>`. It would be consistent with the rest of `engine`. We decided not to force any convention (see [comment](https://github.com/ckeditor/ckeditor5/issues/11719#issuecomment-1148731592)), so it is up to you.